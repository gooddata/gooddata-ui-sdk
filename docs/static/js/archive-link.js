document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});