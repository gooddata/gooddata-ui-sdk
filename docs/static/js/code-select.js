(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));