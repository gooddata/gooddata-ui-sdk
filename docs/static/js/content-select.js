(function ($) {
    $(document).ready(function () {
        $(".gd-docs-content-select__tab").on("click", setActive);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");

            $(this).addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-content__block")
                .removeClass("active")
                .filter(".content-" + $(this).data("content"))
                .addClass("active");
        }
    }
}(jQuery));
