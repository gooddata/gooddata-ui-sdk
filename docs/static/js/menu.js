$(document).ready(function () {
    var menu = $(".gd-docs-menu"),
        menuActive = $(".gd-docs-menu-page.active");

    if(menuActive.length > 0) {
        menu.scrollTop(menuActive.offset().top - menu.offset().top - parseInt(menu.css("paddingTop")));
    }

    menu.removeClass("init");
});