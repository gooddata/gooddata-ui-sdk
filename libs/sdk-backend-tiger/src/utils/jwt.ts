// (C) 2023 GoodData Corporation

import { invariant } from "ts-invariant";

/**
 * Internal interface that contains JWT fields accessed by the authentication provider.
 * There can be other field (claims) in the token payload as well but these are not important for the
 * provider right now.
 */
export interface IJsonWebTokenPayload {
    /**
     * Number of seconds since the start of epoch when the token will expire.
     */
    exp: number;
    /**
     * Subject to which the token grants the access (ID of the Tiger user in our case).
     */
    sub: string;
}

/**
 * Decode provided JWT payload
 *
 * @param jwt - JWT that should be decoded. Expected JWT structure is "header.payload.signature".
 *
 * @returns object with claims decoded from provided JWT payload
 *
 * @throws when provided jwt is empty
 */
export const decodeJwtPayload = (jwt: string): IJsonWebTokenPayload => {
    invariant(jwt, "The provided token is empty");
    const base64UrlPayload = jwt.split(".")[1];
    const base64Payload = base64UrlPayload.replace(/-/g, "+").replace(/_/g, "/");
    const jsonPayload = decodeURIComponent(
        atob(base64Payload)
            .split("")
            .map((ch) => "%" + ("00" + ch.charCodeAt(0).toString(16)).slice(-2))
            .join(""),
    );
    return JSON.parse(jsonPayload);
};

/**
 * Compare subject from the new and previous JWT and compare if the subject is the same for both ("sub" claim)
 *
 * @param newJwt - new JWT
 * @param previousJwt - previous JWT
 *
 * @throws when "sub" claim does not equal in both compared JWTs
 */
export const validateJwt = (newJwt: string, previousJwt: string): void => {
    const previousPayload = decodeJwtPayload(previousJwt);
    const payload = decodeJwtPayload(newJwt);
    if (payload.sub !== previousPayload.sub) {
        throw new Error("The new JWT does not belong to the same subject as the previous one.");
    }
};

/**
 * Compute number of milliseconds till provided JWT expiration from the current time.
 * The returned value can be used for setTimeout function.
 *
 * @param jwt - JWT from which the expiration timestamp is extracted
 * @param secondsBeforeTokenExpirationToCallReminder - number of seconds subtracted from the final timeout
 * @param currentTimeInMilliseconds - current time in milliseconds
 *
 * @returns number of milliseconds between now and expiration of provided JWT, subtracted by provided number of seconds
 */
export const computeExpirationReminderTimeout = (
    jwt: string,
    secondsBeforeTokenExpirationToCallReminder: number,
    currentTimeInMilliseconds: number,
): number => {
    const payload = decodeJwtPayload(jwt);
    const millisecondsToReminder = secondsBeforeTokenExpirationToCallReminder * 1000;
    const epochMillisecondsToExpiration = payload.exp * 1000;
    return epochMillisecondsToExpiration - currentTimeInMilliseconds - millisecondsToReminder;
};
