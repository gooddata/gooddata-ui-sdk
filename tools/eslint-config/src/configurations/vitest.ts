// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
