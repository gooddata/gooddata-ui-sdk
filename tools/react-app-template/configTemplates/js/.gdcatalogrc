{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}