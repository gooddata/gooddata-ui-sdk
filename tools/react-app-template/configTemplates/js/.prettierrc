{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
