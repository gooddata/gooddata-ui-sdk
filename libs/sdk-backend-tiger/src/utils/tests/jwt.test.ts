// (C) 2023 GoodData Corporation

import { describe, expect, it } from "vitest";

import { computeExpirationReminderTimeout, validateJwt, decodeJwtPayload } from "../jwt.js";

const jwt1 =
    "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VMm9wT2xnRzV0UFlKNm9pSk9VczFUMXB1OCJ9.eyJzdWIiOiJwZXRyLmRvbGVqc2kiLCJuYW1lIjoiUGV0ciBEb2xlanNpIiwiaWF0IjoxNjg5MTY0MTk3LCJqdGkiOiI3N2ZkZDI4OS00MDkyLTRkMzQtOThmNC00ZDRkN2YyNWZjMTQiLCJleHAiOjE2ODkxNjQyMjd9.keJ9yKu7DuIMM7Z3N8Aaw44PRHqS4NI2FSYcuvABlDz-j2dw3Nc5Djt5qczr_Yfs0HahmstZugoUoszzrc6S9UcJLaoeljn9Ftouwuqr8ArzVurdk51xQDK2u4fMx-S96J93MCbBg2onRoE0WHxzrENwt82QwVXD44dObqYXEzFFh3E5z65F3XMUeuI_BLwAsrdTVwqA9ZyYaXIqOFXFaFZwrXK2EVjg_NISLx0I-8g6yzpVrS5d32ezhWM0aRiy-O2P-b0SC1Eu_BX9FXiSqNvZnkRiERFCfYYYQrf9TvobC7UesBj1bYfgNgHHP18wd2Zru1-V5mvdEnAF19omAw";
const jwt2 =
    "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VMm9wT2xnRzV0UFlKNm9pSk9VczFUMXB1OCJ9.eyJzdWIiOiJwZXRyLmRvbGVqc2kiLCJuYW1lIjoiUGV0ciBEb2xlanNpIiwiaWF0IjoxNjg5MTY0Mzg5LCJqdGkiOiIxNGNkMTVlYi1lYjgwLTQ1ZGUtOTk2OS1kMjEzZjdiZjVkY2UiLCJleHAiOjE2ODkxNjQ0MTl9.MpIMB1d5JYaxojkrtswN4eaGlH3FKsISnqb7f8iETWDtv6D4J6cJJIuJg34q9v7C8--mAEVVbUyDI4U6jlP4TJ0P2tcvkTmFEjXWLrweOum_E_MzcxvUekksV-k9kIfG6dv0Q11uQn_g7tV40w0GLaPxFPZvOvvNfsXTs1OQlsn_6QpL2-XPQuNE-BE2e9OogOzA_h3A-EkiJCLHn-sn1aNgiRZEeiMweMLuQ2KhMpEFZIrHvA8QHG2QXWtdeeQ05PD6-LnghZRDBLWXxuz7x5JSBapC90R3D4acJ5639Krwvb2x4bKc8sJOm9TPBZ79lKdqKKl4YboNPeX1dDzbTg";
const jwt3 =
    "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VMm9wT2xnRzV0UFlKNm9pSk9VczFUMXB1OCJ9.eyJzdWIiOiJtb3J0IiwibmFtZSI6IlBldHIgRG9sZWpzaSIsImlhdCI6MTY4OTE2NDQxMCwianRpIjoiOTNmMWI5OGItOWFmNC00OGZiLTgwMGEtN2VlNThhYTA0NjFhIiwiZXhwIjoxNjg5MTY0NDQwfQ.HFEa5Xx5AWaSV16fT7oFWeovKV1WRrqdsyOfQpE8vR-BfNnEVMJ68vO5emM02O-fPvZAjsvwMUjV65YRondSSWo-OKVCVL1DMHzytwA_mAUJyQhXv8WqFa8bdeVrKiV1S3w51p2JmKFLfmBz_EVX5HTSzFlkEXa8MYttiiO_hPOJ3b_F6sv-wiVgTrGDjSn3JJ0tbysC_DCHhc0q-z1LL3t_20lkpOa_1fOGahVOwkXgyn66Fgm7ipTTtz9kxvN2IBy-lVIv-ixqQBW9PlMUosPdn4hMGRz-93qZmByzL3XsdhpfdORr6poQF1lpXxImQg03gXAREoh_vduAF2GK3g";

describe("jwt", () => {
    describe("decodeJwtPayload", () => {
        it("should throw when token is not provided", () => {
            expect(() => decodeJwtPayload("")).toThrowError();
        });

        it("should parse payload of JWT", () => {
            expect(decodeJwtPayload(jwt1)).toMatchSnapshot();
        });
    });

    describe("validateJwt", () => {
        it("should not throw when two different JWTs from the same subject are provided", () => {
            expect(() => validateJwt(jwt1, jwt2)).not.toThrowError();
        });

        it("should throw when two different JWTs for the different subjects are provided", () => {
            expect(() => validateJwt(jwt2, jwt3)).toThrowError();
        });
    });

    describe("computeExpirationReminderTimeout", () => {
        it("should correctly count the time to expiration reminder in milliseconds", () => {
            const timeout = computeExpirationReminderTimeout(jwt1, 6, 1689164199625);
            expect(timeout).toBe(21375);
        });
    });
});
