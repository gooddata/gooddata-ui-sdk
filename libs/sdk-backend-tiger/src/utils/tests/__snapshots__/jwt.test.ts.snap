// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jwt > decodeJwtPayload > should parse payload of JWT 1`] = `
{
  "exp": 1689164227,
  "iat": 1689164197,
  "jti": "77fdd289-4092-4d34-98f4-4d4d7f25fc14",
  "name": "Petr Dolejsi",
  "sub": "petr.dolejsi",
}
`;
